packages/vite/src/node/plugins/asset.ts

```diff
@@ -222,10 +222,31 @@ async function fileToBuiltUrl(
     if (!map.has(contentHash)) {
       const basename = path.basename(file)
       const ext = path.extname(basename)
-      const fileName = path.posix.join(
+      let fileName = path.posix.join(
         config.build.assetsDir,
         `${basename.slice(0, -ext.length)}.${contentHash}${ext}`
       )
+
+      const { output } = config.build?.rollupOptions ?? {}
+      // Only the object format is currently considered here.
+      if (output && !Array.isArray(output)) {
+        const assetFileNames = output.assetFileNames
+        // e.g assetFileNames: dir/[name].[ext]
+        if (typeof assetFileNames === 'string') {
+          fileName = assetFileNames
+            .replace(/\[name\]/, basename.slice(0, -ext.length))
+            .replace(/\[ext\]/, ext)
```

Comment:
`ext` can have a dot (as it is the result of path.extname)
```suggestion
            .replace(/\[ext\]/, ext.trimLeft('.'))
```

Reply:
Sorry @anncwb, I didn't test my suggestion. You could use something like `ext.replace(/^\./,'')`